src/npyffi/array.rs

```diff
@@ -1,361 +1,338 @@
 //! Low-Level binding for [Array API](https://docs.scipy.org/doc/numpy/reference/c-api.array.html)
-
 use libc::FILE;
+use pyo3::ffi::{self, PyObject, PyTypeObject};
 use std::ops::Deref;
 use std::os::raw::*;
-use std::ptr::null_mut;
-
-use pyo3::ffi;
-use pyo3::ffi::{PyObject, PyTypeObject};
-use pyo3::{ObjectProtocol, PyModule, PyResult, Python, ToPyPointer};
+use std::ptr;
 
 use npyffi::*;
 
-/// Low-Level binding for [Array API](https://docs.scipy.org/doc/numpy/reference/c-api.array.html)
-///
-/// Most of the Array APIs are exposed as related functions of this module object.
-/// Some APIs (including most accessor) in the above URL are not implemented
-/// since they are defined as C macro, and cannot be called from rust.
-/// Some of these are implemented on the high-level interface as a Rust function.
-pub struct PyArrayModule<'py> {
-    numpy: &'py PyModule,
-    api: *const *const c_void,
-}
-
-#[allow(non_upper_case_globals)]
-pub(crate) static mut PyArray_Type_Ptr: *mut PyTypeObject = null_mut();
+pub(crate) const MOD_NAME: &str = "numpy.core.multiarray";
+const CAPSULE_NAME: &str = "_ARRAY_API";
 
-impl<'py> Deref for PyArrayModule<'py> {
-    type Target = PyModule;
-    fn deref(&self) -> &Self::Target {
-        self.numpy
-    }
-}
+/// A global variable which stores a ['capsule'](https://docs.python.org/3/c-api/capsule.html)
+/// pointer to [Numpy Array API](https://docs.scipy.org/doc/numpy/reference/c-api.array.html).
+///
+/// You can acceess raw c APIs via this variable and its Deref implementation.
+///
+/// See [PyArrayAPI](struct.PyArrayAPI.html) for what methods you can use via this variable.
+///
+/// # Example
+/// ```
+/// # extern crate pyo3; extern crate numpy; fn main() {
+/// use numpy::{PyArray, npyffi::types::NPY_SORTKIND, PY_ARRAY_API};
+/// use pyo3::Python;
+/// let gil = Python::acquire_gil();
+/// let array: &PyArray<i32> = PyArray::arange(gil.python(), 2.0, 5.0, 1.0);
+/// array.as_slice_mut().unwrap().swap(0, 1);
+/// assert_eq!(array.as_slice().unwrap(), &[3, 2, 4]);
+/// unsafe {
+///     PY_ARRAY_API.PyArray_Sort(array.as_array_ptr(), 0, NPY_SORTKIND::NPY_QUICKSORT);
+/// }
+/// assert_eq!(array.as_slice().unwrap(), &[2, 3, 4])
+/// # }
+/// ```
+pub static PY_ARRAY_API: PyArrayAPI = PyArrayAPI {
+    __private_field: (),
+};
 
-/// Define Array API in PyArrayModule
-macro_rules! pyarray_api {
-    [ $offset:expr; $fname:ident ( $($arg:ident : $t:ty),* ) $( -> $ret:ty )* ] => {
-#[allow(non_snake_case)]
-pub unsafe fn $fname(&self, $($arg : $t), *) $( -> $ret )* {
-    let fptr = self.api.offset($offset) as (*const extern fn ($($arg : $t), *) $( -> $ret )* );
-    (*fptr)($($arg), *)
+///
+pub struct PyArrayAPI {
+    __private_field: (),
```

Comment:
Can we remove this field?

Reply:
I think we need this, to prevent  users from constructing this struct in external crates.

Reply:
I see. thanks